src/shrpx.cc

```diff
@@ -386,6 +403,9 @@ void exec_binary() {
     if (pid == -1) {
       auto error = errno;
       LOG(ERROR) << "fork() failed errno=" << error;
+    } else {
+      // update PID tracking information in systemd
+      shrpx_sd_notifyf(0, "MAINPID=%s\n", util::utos(pid).c_str());
```

Comment:
Yes it is required here. systemd only accepts sd_notify() messages from main PID of the service (this could be relaxed in configuration, but it is more secure that way).  Thus, in this place we are still a main PID, but we started *new* master process and we are basically telling systemd "hey, this is new master process of this service; please listen for notification from this new PID).

We cannot rely on message in line 1279, because it would be ignored by systemd - it would come from (new master) PID, which is not known for systemd as master PID.
Line 1279 works during normal startup because then we are only process in a service, thus we are a MAINPID.

I hope that is clear.